rfcs/converged-implementation-patterns.md

```diff
@@ -0,0 +1,171 @@
+-->
+
+Many new developers might have questions on the patterns that we follow to write a React component. This RFC attempts to commit and possibly improve our code patterns in writing so that there might be a reference for new (and old) members of the team.
+
+## Detailed Design or Proposal
+
+<!-- This is the bulk of the RFC. Explain the proposal or design in enough detail for the inteded audience to understand. -->
+
+### Hook based architecture.
+
+Generally a component will have the following different files. Let us consider a `Sample` component
+
+- Sample.tsx
+- useSample.ts
+- useSampleClasses/useSampleStyles.ts
+- renderSample.tsx
+- Sample.types.ts
+
+#### renderSample.tsx
+
+Renders the correct JSX output of the component and its slots given the correct state.
+
+```typescript=
+export const renderSample = (state: SampleState) => {
+  const { slots, slotProps } = getSlots(state);
+
+  return (
+    <slots.root {...slotProps.root}>
+      <slot.otherSlot {...slotProps.otherSlot} />
+    </slots.root>
+  );
+};
+```
+
+#### useSample.ts
+
+Accepts the component props and handles and internal state that the component might need.
+
+`State` here can be pretty broad, you could also consume context or create effects. This hook should be what the component relies on to function/render
+
+```typescript=
+/**
+ * Defines the different slots that can be rendered in this component
+ *
+ * This means that an icon can be rendered 'somewhere'
+ */
+export const sampleShorthandProps = ['icon'];
+
+// Creates a helper function to merge props while respecting slot props
+const mergeProps = makeMergeProps({ deepMerge: sampleShorthandProps });
+
+/**
+ * @parameter props -> these are normal React props for JSX components
+ * @parameter ref -> In case someone wants a ref to the root DOM element
+ * @parameter defaultProps -> safe default props
+ */
+export const useSample = (props: SampleProps, ref: React.Ref<HTMLElement>, defaultProps?: SampleProps) => {
+  const resolvedRef = ref || React.useRef();
+  const { checkedValues, onCheckedValuesChange } = useMenuListContext();
+
+  // merges the props we declare internally and what is passed in
+  // by a consumer
+  const mergedProps = mergeProps(
+    {
+      ref: resolvedRef,
+      as: 'div',
+      icon: { as: 'span' },
+    },
+    defaultProps,
+    resolveShorthandProps(props, sampleShorthandProps),
+  );
+
+  if (checkedValues || onCheckedValuesChange) {
+    mergedProps.hasCheckMark = true;
+  }
+
+  // Anything else the component needs to manage its own state and render
+  [someState, setSomeState] = React.useState();
+  const { contextValue } = React.useContext();
+  React.useEffect(...);
+
+  // make an 'uber' state
+  const state = { someState, contextValue, ...mergedProps}
+
+  return state;
+};
+```
+
+#### useSampleClasses/useSampleStyles.ts
+
+Hook that accepts state and applies classnames to props and any shorthand slot props to style the component and its slots
```

Comment:
Are you going to expand on this one with a code snippet?

Reply:
our styling has been/is in a state of change, I _**think**_  I can include `make-styles` but all current converged components in the repo use the older `makeVariants`

@layershifter  @miroslavstastny  am I OK to go with `makestyles` here ?

Reply:
AFAIR it was agreed that we're gonna go with `makeStyles`, go go go 👍

Reply:
Personally, I think we should slowly converge (pun not intended) towards `makeStyles`, or at least, that's what I intend to incorporate into `react-image`.

Reply:
Agree on an example of using make-styles here would be good. 

It would be reasonable to ask @miroslavstastny or @layershifter to add it in a subsequent RFC (if we go this route, I would recommend we use an issue in Feb milestone to track the followup).

Reply:
I've added a sample usage of `makeStyles` that I'm currently using, from what I understand @miroslavstastny and @layershifter  are also going this route